arangod/RocksDBEngine/RocksDBSettingsManager.cpp

```diff
@@ -115,183 +110,198 @@ void RocksDBSettingsManager::retrieveInitialValues() {
+      TRI_voc_tick_t dbid = pair.first;
+      DataSourceId cid = pair.second;
+      TRI_vocbase_t* vocbase = dbfeature.useDatabase(dbid);
+      if (!vocbase) {
+        continue;
+      }
+      TRI_ASSERT(!vocbase->isDangling());
+      auto sg = arangodb::scopeGuard([&]() noexcept { vocbase->release(); });
 
-    LOG_TOPIC("afb17", TRACE, Logger::ENGINES)
-        << "syncing metadata for collection '" << coll->name() << "'";
+      std::shared_ptr<LogicalCollection> coll;
+      try {
+        coll = vocbase->useCollection(cid, /*checkPermissions*/ false);
+      } catch (...) {
+        // will fail if collection does not exist
+      }
+      // Collections which are marked as isAStub are not allowed to have
+      // physicalCollections. Therefore, we cannot continue serializing in that
+      // case.
+      if (!coll || coll->isAStub()) {
+        continue;
+      }
+      auto sg2 = arangodb::scopeGuard(
+          [&]() noexcept { vocbase->releaseCollection(coll.get()); });
 
-    // clear our scratch buffer for this round
-    _scratch.clear();
+      LOG_TOPIC("afb17", TRACE, Logger::ENGINES)
+          << "syncing metadata for collection '" << coll->name() << "'";
 
-    auto* rcoll = static_cast<RocksDBCollection*>(coll->getPhysical());
-    rocksdb::SequenceNumber appliedSeq = maxSeqNr;
-    Result res = rcoll->meta().serializeMeta(batch, *coll, force, _tmpBuilder,
-                                             appliedSeq, _scratch);
-    minSeqNr = std::min(minSeqNr, appliedSeq);
+      // clear our scratch buffers for this round
+      _scratch.clear();
+      _tmpBuilder.clear();
+      batch.Clear();
 
-    const std::string err = "could not sync metadata for collection '";
-    if (res.fail()) {
-      LOG_TOPIC("1038d", WARN, Logger::ENGINES) << err << coll->name() << "'";
-      return res;
+      auto* rcoll = static_cast<RocksDBCollection*>(coll->getPhysical());
+      rocksdb::SequenceNumber appliedSeq = maxSeqNr;
+      Result res = rcoll->meta().serializeMeta(batch, *coll, force, _tmpBuilder,
+                                               appliedSeq, _scratch);
+
+      if (res.ok() && batch.Count() > 0) {
+        didWork = true;
+
+        auto s = _db->Write(wo, &batch);
+        if (!s.ok()) {
+          res.reset(rocksutils::convertStatus(s));
+        }
+      }
+
+      if (res.fail()) {
+        LOG_TOPIC("afa17", WARN, Logger::ENGINES)
+            << "could not sync metadata for collection '" << coll->name()
+            << "'";
+        return res;
+      }
+
+      minSeqNr = std::min(minSeqNr, appliedSeq);
     }
 
-    if (batch.Count() > 0) {
-      auto s = _db->Write(wo, &batch);
-      if (!s.ok()) {
-        LOG_TOPIC("afa17", WARN, Logger::ENGINES) << err << coll->name() << "'";
-        return rocksutils::convertStatus(s);
+    if (_scratch.capacity() >= 32 * 1024 * 1024) {
+      // much data in _scratch, let's shrink it to save memory
+      _scratch.clear();
```

Comment:
Maybe resize back to initial "128 * 1024" here? To avoid deallocation-alocation cycles?

Reply:
Also why clear + shrink_to_fit(can do nothing) between `.reset(0);` or  ` = {};`?